File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/model/DeviceModels.kt
package dev.bota.sdk.model

import java.time.Instant

public sealed interface WireValue<out T> {
    public val unknownRawValue: ULong?

    public data class Known<T>(public val value: T) : WireValue<T> {
        override val unknownRawValue: ULong? = null
    }

    public data class Unknown(public val rawValue: ULong) : WireValue<Nothing> {
        override val unknownRawValue: ULong = rawValue
    }
}

public sealed interface DeviceType {
    public data object BotaPin : DeviceType
    public data object BotaPin4G : DeviceType
    public data object BotaNote : DeviceType
    public data class Unknown(public val rawValue: UByte) : DeviceType
}

public sealed interface PairingState {
    public data object Unpaired : PairingState
    public data object Pairing : PairingState
    public data object Paired : PairingState
    public data object Error : PairingState
    public data class Unknown(public val rawValue: UByte) : PairingState
}

public enum class ConnectionState {
    Disconnected,
    Connecting,
    Bonding,
    Discovering,
    Connected,
    Disconnecting,
}

public enum class DeviceState {
    Idle,
    Recording,
    Syncing,
    Uploading,
    Charging,
    LowBattery,
    StorageFull,
    Error,
}

public enum class LteStatus {
    Off,
    Searching,
    Registered,
    Connected,
    Denied,
    NoSim,
    Error,
    LowVoltage,
    Disabled,
}

public enum class WifiRadioStatus {
    Off,
    Scanning,
    Connecting,
    Connected,
    ConnectFailed,
    NoCredentials,
    Disabled,
    Error,
}

public data class DeviceFlags(
    public val charging: Boolean,
    public val lowBattery: Boolean,
    public val storageFull: Boolean,
    public val wifiConnected: Boolean,
    public val lteConnected: Boolean,
    public val syncActive: Boolean,
)

public data class ModemInfo(
    public val imei: String? = null,
    public val iccid: String? = null,
    public val operator: String? = null,
    public val rat: String? = null,
    public val band: String? = null,
    public val apn: String? = null,
    public val simStatus: String? = null,
    public val csq: Int? = null,
    public val ipAddress: String? = null,
    public val modemVoltage: Int? = null,
    public val modemFirmware: String? = null,
    public val roaming: Boolean? = null,
)

public data class DeviceStatus(
    public val batteryLevel: Int,
    public val batteryMv: Int? = null,
    public val storageTotalMb: Int,
    public val storageUsedMb: Int,
    public val state: WireValue<DeviceState>,
    public val pendingRecordings: Int,
    public val lastTimeSyncAt: Instant?,
    public val signalStrength: Int = 0,
    public val flags: DeviceFlags,
    public val timestamp: UInt,
    public val lteStatus: WireValue<LteStatus>,
    public val lteSignalQuality: Int? = null,
    public val wifiStatus: WireValue<WifiRadioStatus>? = null,
    public val modemInfo: ModemInfo? = null,
)

public class DiscoveredDevice(
    public val id: String,
    public val name: String? = null,
    public val deviceType: DeviceType? = null,
    public val firmwareVersion: String? = null,
    public val macAddress: String? = null,
    public val pairingState: PairingState? = null,
    public val rssi: Int,
    manufacturerData: ByteArray? = null,
    public val discoveredAt: Instant = Instant.now(),
) {
    private val storedManufacturerData: ByteArray? = manufacturerData?.copyOf()
    public val manufacturerData: ByteArray? get() = storedManufacturerData?.copyOf()

    override fun equals(other: Any?): Boolean = other is DiscoveredDevice &&
        id == other.id && name == other.name && deviceType == other.deviceType &&
        firmwareVersion == other.firmwareVersion && macAddress == other.macAddress &&
        pairingState == other.pairingState && rssi == other.rssi &&
        storedManufacturerData.contentEqualsNullable(other.storedManufacturerData) &&
        discoveredAt == other.discoveredAt

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + (name?.hashCode() ?: 0)
        result = 31 * result + (deviceType?.hashCode() ?: 0)
        result = 31 * result + (firmwareVersion?.hashCode() ?: 0)
        result = 31 * result + (macAddress?.hashCode() ?: 0)
        result = 31 * result + (pairingState?.hashCode() ?: 0)
        result = 31 * result + rssi
        result = 31 * result + (storedManufacturerData?.contentHashCode() ?: 0)
        return 31 * result + discoveredAt.hashCode()
    }
}

public data class ConnectedDevice(
    public val id: String,
    public val serialNumber: String,
    public val deviceType: DeviceType,
    public val firmwareVersion: String,
    public val hardwareRevision: String? = null,
    public val isProvisioned: Boolean,
    public val connectionState: ConnectionState,
    public val mtu: Int,
)

internal fun ByteArray?.contentEqualsNullable(other: ByteArray?): Boolean = when {
    this == null -> other == null
    other == null -> false
    else -> contentEquals(other)
}

File: platforms/apple/Sources/BotaAppSDK/Models/DeviceModels.swift
import Foundation

public enum WireValue<Known: Equatable & Sendable>: Equatable, Sendable {
    case known(Known)
    case unknown(UInt64)
}

public enum DeviceType: Equatable, Sendable {
    case botaPin
    case botaPin4G
    case botaNote
    case unknown(UInt8)
}

public enum PairingState: Equatable, Sendable {
    case unpaired
    case pairing
    case paired
    case error
    case unknown(UInt8)
}

public enum ConnectionState: Equatable, Sendable {
    case disconnected
    case connecting
    case bonding
    case discovering
    case connected
    case disconnecting
}

public enum DeviceState: Equatable, Sendable {
    case idle
    case recording
    case syncing
    case uploading
    case charging
    case lowBattery
    case storageFull
    case error
}

public enum LteStatus: Equatable, Sendable {
    case off
    case searching
    case registered
    case connected
    case denied
    case noSim
    case error
    case lowVoltage
    case disabled
}

public enum WifiRadioStatus: Equatable, Sendable {
    case off
    case scanning
    case connecting
    case connected
    case connectFailed
    case noCredentials
    case disabled
    case error
}

public struct DeviceFlags: Equatable, Sendable {
    public var charging: Bool
    public var lowBattery: Bool
    public var storageFull: Bool
    public var wifiConnected: Bool
    public var lteConnected: Bool
    public var syncActive: Bool

    public init(
        charging: Bool,
        lowBattery: Bool,
        storageFull: Bool,
        wifiConnected: Bool,
        lteConnected: Bool,
        syncActive: Bool
    ) {
        self.charging = charging
        self.lowBattery = lowBattery
        self.storageFull = storageFull
        self.wifiConnected = wifiConnected
        self.lteConnected = lteConnected
        self.syncActive = syncActive
    }
}

public struct ModemInfo: Equatable, Sendable {
    public var imei: String?
    public var iccid: String?
    public var `operator`: String?
    public var rat: String?
    public var band: String?
    public var apn: String?
    public var simStatus: String?
    public var csq: Int?
    public var ipAddress: String?
    public var modemVoltage: Int?
    public var modemFirmware: String?
    public var roaming: Bool?

    public init(
        imei: String? = nil,
        iccid: String? = nil,
        operator: String? = nil,
        rat: String? = nil,
        band: String? = nil,
        apn: String? = nil,
        simStatus: String? = nil,
        csq: Int? = nil,
        ipAddress: String? = nil,
        modemVoltage: Int? = nil,
        modemFirmware: String? = nil,
        roaming: Bool? = nil
    ) {
        self.imei = imei
        self.iccid = iccid
        self.operator = `operator`
        self.rat = rat
        self.band = band
        self.apn = apn
        self.simStatus = simStatus
        self.csq = csq
        self.ipAddress = ipAddress
        self.modemVoltage = modemVoltage
        self.modemFirmware = modemFirmware
        self.roaming = roaming
    }
}

public struct DeviceStatus: Equatable, Sendable {
    public var batteryLevel: Int
    public var batteryMv: Int?
    public var storageTotalMb: Int
    public var storageUsedMb: Int
    public var state: WireValue<DeviceState>
    public var pendingRecordings: Int
    public var lastTimeSyncAt: Date?
    public var signalStrength: Int
    public var flags: DeviceFlags
    public var timestamp: UInt32
    public var lteStatus: WireValue<LteStatus>
    public var lteSignalQuality: Int?
    public var wifiStatus: WireValue<WifiRadioStatus>?
    public var modemInfo: ModemInfo?

    public init(
        batteryLevel: Int,
        batteryMv: Int? = nil,
        storageTotalMb: Int,
        storageUsedMb: Int,
        state: WireValue<DeviceState>,
        pendingRecordings: Int,
        lastTimeSyncAt: Date?,
        signalStrength: Int = 0,
        flags: DeviceFlags,
        timestamp: UInt32,
        lteStatus: WireValue<LteStatus>,
        lteSignalQuality: Int? = nil,
        wifiStatus: WireValue<WifiRadioStatus>? = nil,
        modemInfo: ModemInfo? = nil
    ) {
        self.batteryLevel = batteryLevel
        self.batteryMv = batteryMv
        self.storageTotalMb = storageTotalMb
        self.storageUsedMb = storageUsedMb
        self.state = state
        self.pendingRecordings = pendingRecordings
        self.lastTimeSyncAt = lastTimeSyncAt
        self.signalStrength = signalStrength
        self.flags = flags
        self.timestamp = timestamp
        self.lteStatus = lteStatus
        self.lteSignalQuality = lteSignalQuality
        self.wifiStatus = wifiStatus
        self.modemInfo = modemInfo
    }
}

public struct DiscoveredDevice: Equatable, Sendable {
    public var id: String
    public var name: String?
    public var deviceType: DeviceType?
    public var firmwareVersion: String?
    public var macAddress: String?
    public var pairingState: PairingState?
    public var rssi: Int
    public var manufacturerData: Data?
    public var discoveredAt: Date

    public init(
        id: String,
        name: String? = nil,
        deviceType: DeviceType? = nil,
        firmwareVersion: String? = nil,
        macAddress: String? = nil,
        pairingState: PairingState? = nil,
        rssi: Int,
        manufacturerData: Data? = nil,
        discoveredAt: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.deviceType = deviceType
        self.firmwareVersion = firmwareVersion
        self.macAddress = macAddress
        self.pairingState = pairingState
        self.rssi = rssi
        self.manufacturerData = manufacturerData
        self.discoveredAt = discoveredAt
    }
}

public struct ConnectedDevice: Equatable, Sendable {
    public var id: String
    public var serialNumber: String
    public var deviceType: DeviceType
    public var firmwareVersion: String
    public var hardwareRevision: String?
    public var isProvisioned: Bool
    public var connectionState: ConnectionState
    public var mtu: Int

    public init(
        id: String,
        serialNumber: String,
        deviceType: DeviceType,
        firmwareVersion: String,
        hardwareRevision: String? = nil,
        isProvisioned: Bool,
        connectionState: ConnectionState,
        mtu: Int
    ) {
        self.id = id
        self.serialNumber = serialNumber
        self.deviceType = deviceType
        self.firmwareVersion = firmwareVersion
        self.hardwareRevision = hardwareRevision
        self.isProvisioned = isProvisioned
        self.connectionState = connectionState
        self.mtu = mtu
    }
}
